activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/persistence/entity/ExecutionEntityManagerImpl.java

```diff
@@ -419,6 +425,14 @@ protected void deleteProcessInstanceCascade(ExecutionEntity execution, String de
       return;
     }
 
+      // Execute execution listeners for process end.
+      Process process = ProcessDefinitionUtil.getProcess(processInstanceExecutionEntity.getProcessDefinitionId());
+      if (CollectionUtil.isNotEmpty(process.getExecutionListeners())) {
+          executeExecutionListeners(process,
+              processInstanceExecutionEntity,
+              ExecutionListener.EVENTNAME_END);
+      }
```

Comment:
Suggestion for future developments: prefer to extract a method with a meaningful name instead of using a comment to tell what the code block does.

Reply:
Thanks for the input, it's a good contribution for CI: Continuous Improvement 👍